### DISCUSSION TITLE
Bring Curve2D editing plugin to exported variables in any custom nodes

### DISCUSSION THREAD

**Discussion by Exerionius** [godotengine/godot-proposals#3997] [Editor]
Hello everyone.

There is a great built-in "plugin" for editing Curve2D (think of bezier curves) in the editor: https://github.com/godotengine/godot/blob/master/editor/plugins/path_2d_editor_plugin.cpp
![image](https://user-images.githubusercontent.com/13285909/154745274-42ebd28b-d3d9-4c4c-adac-3a190cf234ac.png)

But this plugin works only on the Path2D node. If the user exports a Curve2D variable in their custom node, the editing functionality is not there:

``` GDScript
tool
class_name TestNode
extends Node2D

@export  var curve : Curve2D
```

If the plugin is already there, why not to enable it for any Curve2D exported variable?

Currently I have to create a temporal Path2D node just for editing a curve, saving it into a file resource and loading back into my node, because it's impossible to edit it on the node itself.

Imagine if you have something like this:

``` GDScript
@export var ammo : int
```

And you need to create a separate special node just to edit this value before loading it back into your exported variable every time you need to adjust it in the editor.

**Comment by Xrayez** [comment]
I'd say this is a major limitation with editor plugins in general. Or rather, those plugins that require Godot's canvas item editor.

The same issue is present with `NavigationPolygon`, `OccluderPolygon2D`, `Shape[2D]`, and likely others.

For example, a [`VisualShape2D`](https://goost.readthedocs.io/en/latest/classes/class_visualshape2d.html) that you can find in [Goost](https://goostengine.github.io/) cannot edit the `Shape2D` resource directly, unfortunately. Reusing the same editor plugin for custom nodes is currently impossible, and reimplementing those plugins is frankly impractical. See also #1126, #1157.

**Comment by Exerionius** [comment]
The question is - how difficult it is to reuse it, considering that the functionality is already there.
I'm not familiar with Godot sources and my C++ is rather rusty, so I can't really tell myself.

**Reply by Xrayez** [reply]
First, the plugin has to be exposed to scripting. You can verify this is not the case currently. The only specific plugin I see there exposed is `AnimationTrackEditPlugin`, and probably a few others.

Second, the source code directly relies on whether a node is selected. For this to work directly on a resource, it has to `handles()` the resource directly (`Curve2D`). So this requires C++ core modifications in either case. The rest of functionality probably doesn't need to be modified as much. Perhaps there will need to be created a new plugin that would extend the base, so that the plugin could edit both the node containing the resource, and the resource itself.

**Comment by mutualtape** [comment]
Had the same need and came up with this workaround in my custom node plugin that works well enough for a smooth workflow. It creates a Path2D on the fly (I use the method to add other nodes like collision etc. too) and adds a listener on changes to the curve. 
```
@tool
@onready var curve_node: Path2D = lazy_add_to_scene("Curve", Path2D)
	
func lazy_add_to_scene(name, clazz: Variant):
	var node
	if(has_node(name)): node = get_node(name)
	else:
		node = clazz.new()
		node.name = name
		if("curve" in node): node.curve = Curve2D.new()
		add_child(node)
		node.owner = get_tree().edited_scene_root
	return node	
	
func _ready():
	if(Engine.is_editor_hint()):
		curve_node.curve.changed.connect(func(): 
			# copy curve to other persistent nodes:
			a_collision_node.polygon = curve_node.curve.get_baked_points()
		)
```
